src/core/colorspace_utils.js

```diff
@@ -0,0 +1,349 @@
+      globalColorSpaceCache.set(/* name = */ null, csRef, parsedCS);
+    }
+    return parsedCS;
+  }
+
+  static #parse(cs, options) {
+    const { xref, resources, pdfFunctionFactory } = options;
+
+    cs = xref.fetchIfRef(cs);
+    if (cs instanceof Name) {
+      switch (cs.name) {
+        case "G":
+        case "DeviceGray":
+          return this.singletons.gray;
+        case "RGB":
+        case "DeviceRGB":
+          return this.singletons.rgb;
+        case "DeviceRGBA":
+          return this.singletons.rgba;
+        case "CMYK":
+        case "DeviceCMYK":
+          return this.singletons.cmyk;
+        case "Pattern":
+          return new PatternCS(/* baseCS = */ null);
+        default:
+          if (resources instanceof Dict) {
+            const colorSpaces = resources.get("ColorSpace");
+            if (colorSpaces instanceof Dict) {
+              const resourcesCS = colorSpaces.get(cs.name);
+              if (resourcesCS) {
+                if (resourcesCS instanceof Name) {
+                  return this.#parse(resourcesCS, options);
+                }
+                cs = resourcesCS;
+                break;
+              }
+            }
+          }
+          // Fallback to the default gray color space.
+          warn(`Unrecognized ColorSpace: ${cs.name}`);
+          return this.singletons.gray;
+      }
+    }
+    if (Array.isArray(cs)) {
+      const mode = xref.fetchIfRef(cs[0]).name;
+      let params, numComps, baseCS, whitePoint, blackPoint, gamma;
+
+      switch (mode) {
+        case "G":
+        case "DeviceGray":
+          return this.singletons.gray;
+        case "RGB":
+        case "DeviceRGB":
+          return this.singletons.rgb;
+        case "CMYK":
+        case "DeviceCMYK":
+          return this.singletons.cmyk;
+        case "CalGray":
+          params = xref.fetchIfRef(cs[1]);
+          whitePoint = params.getArray("WhitePoint");
+          blackPoint = params.getArray("BlackPoint");
+          gamma = params.get("Gamma");
+          return new CalGrayCS(whitePoint, blackPoint, gamma);
+        case "CalRGB":
+          params = xref.fetchIfRef(cs[1]);
+          whitePoint = params.getArray("WhitePoint");
+          blackPoint = params.getArray("BlackPoint");
+          gamma = params.getArray("Gamma");
+          const matrix = params.getArray("Matrix");
+          return new CalRGBCS(whitePoint, blackPoint, gamma, matrix);
+        case "ICCBased":
+          const { globalColorSpaceCache } = options;
+          const cachedCS = globalColorSpaceCache.getByRef(cs[1]);
+          if (cachedCS) {
+            return cachedCS;
+          }
+
+          const stream = xref.fetchIfRef(cs[1]);
+          const dict = stream.dict;
+          numComps = dict.get("N");
+
+          if (IccColorSpace.isUsable) {
+            try {
+              const iccCS = new IccColorSpace(stream.getBytes(), numComps);
+              globalColorSpaceCache.set(/* name = */ null, cs[1], iccCS);
```

Comment:
... and similarly here as well.
```suggestion
              if (cs[1] instanceof Ref) {
                globalColorSpaceCache.set(/* name = */ null, cs[1], iccCS);
              }
```